pages/common/routers/form.js

```diff
@@ -85,19 +87,24 @@ module.exports = ({
   };
 
   const _process = (req, res, next) => {
-    req.form.values = reduce(req.form.schema, (all, { format }, key) => {
-      const value = trim(req.body[key]);
-      format = format || identity;
-      if (!isUndefined(value)) {
-        all[key] = format(value);
+    req.form.values = reduce(req.form.schema, (all, { format, nullValue }, key) => {
+      let value = req.body[key];
+      if (!value && !isUndefined(nullValue)) {
+        value = nullValue;
       }
+      value = trim(value);
+      format = format || identity;
+      all[key] = format(value);
       return all;
     }, {});
     return process(req, res, next);
   };
 
   const _checkChanged = (req, res, next) => {
-    if (!size(req.form.schema) || some(req.form.values, (value, key) => hasChanged(value, key, req.model))) {
+    const fields = pickBy(req.form.schema, (field, key) => {
+      return field.editable !== false && Object.keys(req.model).includes(key);
+    });
+    if (!size(fields) || some(fields, (field, key) => hasChanged(field, req.form.values[key], req.model[key]))) {
```

Comment:
I think this would be easier to understand if we rolled the `hasChanged` check into the `pickBy` above, so that `fields` only contains the fields in the schema _that have changed_ (and then rename `fields` to `changedFields`)

That condition is pretty wild as it is - 4 consecutive closing brackets is a good indicator of some code that needs a bit of a refactor in my eyes.

Reply:
OWZAT?